### PR TITLE
support Elastic APM for tracing

### PR DESCRIPTION
**Pull request by Wing924**
Signed-off-by: Wing924 <weihe924stephen@gmail.com>

* [] CHANGELOG entry if change is relevant to the end user.

## Changes

- support elastic APM for tracing

## Verification

- setup [elastic APM server](https://www.elastic.co/products/apm)
- config [env variable](https://www.elastic.co/guide/en/apm/agent/go/current/configuration.html) for APM client
- run thanos with `--tracing.config 'type: ELASTIC_APM'`

### CODE REVIEW

**Comment by Wing924** [comment]
Test failed because
```
# cd .; git clone https://git.apache.org/thrift.git /tmp/thanos-go/src/git.apache.org/thrift.git
Cloning into '/tmp/thanos-go/src/git.apache.org/thrift.git'...
fatal: unable to access 'https://git.apache.org/thrift.git/': Failed to connect to git.apache.org port 443: Connection timed out
package github.com/minio/minio
	imports git.apache.org/thrift.git/lib/go/thrift: exit status 128
```

File: go.sum [hunk 1]
```diff
@@ -260,8 +266,11 @@ github.com/hpcloud/tail v1.0.0/go.mod h1:ab1qPbhIpdTxEkNHXyeSf5vhxWSCs/tWer42PpO
 github.com/imdario/mergo v0.3.5/go.mod h1:2EnlNZ0deacrJVfApfmtdGgDfMuh/nq6Ok1EcJh5FfA=
 github.com/influxdata/influxdb v1.7.7/go.mod h1:qZna6X/4elxqT3yI9iZYdZrWWdeFOOprn86kgg4+IzY=
 github.com/jessevdk/go-flags v0.0.0-20180331124232-1c38ed7ad0cc/go.mod h1:4FA24M0QyGHXBuZZK/XkWh8h0e1EYbRYJSGM75WSRxI=
+github.com/jessevdk/go-flags v1.4.0/go.mod h1:4FA24M0QyGHXBuZZK/XkWh8h0e1EYbRYJSGM75WSRxI=
 github.com/jmespath/go-jmespath v0.0.0-20180206201540-c2b33e8439af h1:pmfjZENx5imkbgOkpRUYLnmbU7UEFbjtDA2hxJ1ichM=
 github.com/jmespath/go-jmespath v0.0.0-20180206201540-c2b33e8439af/go.mod h1:Nht3zPeWKUH0NzdCt2Blrr5ys8VGpn0CEB0cQHVjt7k=
+github.com/joeshaw/multierror v0.0.0-20140124173710-69b34d4ec901 h1:rp+c0RAYOWj8l6qbCUTSiRLG/iKnW3K3/QfPPuSsBt4=
+github.com/joeshaw/multierror v0.0.0-20140124173710-69b34d4ec901/go.mod h1:Z86h9688Y0wesXCyonoVr47MasHilkuLMqGhRZ4Hpak=
```
Comment (bwplotka) [inline go.sum]: ..another multi error package to the family ;p 

Nothing we can do, but just funny there are so many libs for a few lines code logic (:

File: pkg/tracing/client/factory.go [hunk 2]
```diff
@@ -5,6 +5,8 @@ import (
 	"io"
 	"strings"
 
+	"github.com/thanos-io/thanos/pkg/tracing/elasticapm"
+
```
Comment (bwplotka) [inline pkg/tracing/client/factory.go]: Can you remove this newline and run `make format`? (limitation of `goimports`)

File: scripts/cfggen/main.go [hunk 3]
```diff
@@ -8,6 +8,8 @@ import (
 	"reflect"
 	"strings"
 
+	"github.com/thanos-io/thanos/pkg/tracing/elasticapm"
+
```
Comment (bwplotka) [inline scripts/cfggen/main.go]: ditto

**Comment by Wing924** [comment]
@bwplotka I fixed the code format. please review it again.

**Comment by bwplotka** [comment]
Awesome. CI is clearly blocked we are working on the fix for: https://github.com/thanos-io/thanos/issues/1483

**Comment by Wing924** [comment]
@bwplotka 
CI is passed. Could you tell me when will you merge or decline it?

**Comment by bwplotka** [comment]
Already approved so.. why decline? (: 

Nice job, thanks!

**Comment by szediktam** [comment]
@Wing924  hi, some questions about elastic_apm config.

I setup a apm server in a k8s cluster, and conntect to elasticsearch, also show "You have correctly setup APM Server" in kibana apm.
I also set a flag in thanos-query as following:
`--tracing.config=type: ELASTIC_APM config:   service_name: "query-trace"   service_version: "1.0.1"   service_environment: "container"   sample_rate: 1`

But there is no apm agent in kibana. something else to do?
thanos-query is a apm agent, or a apm agent need to setup?

thanks.

**Pull request merged**